lifecycle/src/lifecycle_service_client.cpp

```diff
@@ -0,0 +1,161 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+
+#include "lifecycle_msgs/msg/state.hpp"
+#include "lifecycle_msgs/msg/transition.hpp"
+#include "lifecycle_msgs/srv/change_state.hpp"
+#include "lifecycle_msgs/srv/get_state.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+
+// service topics for node-attached services
+static constexpr char const * lifecycle_node = "lc_talker";
+static constexpr char const * node_get_state_topic = "lc_talker__get_state";
+static constexpr char const * node_change_state_topic = "lc_talker__change_state";
+
+class LifecycleServiceClient : public rclcpp::node::Node
+{
+public:
+  explicit LifecycleServiceClient(const std::string & node_name)
+  : rclcpp::node::Node(node_name)
+  {}
+
+  void
+  init()
+  {
+    // Service client pointing to each individual service
+    client_get_state_ = this->create_client<lifecycle_msgs::srv::GetState>(
+      node_get_state_topic);
+    client_change_state_ = this->create_client<lifecycle_msgs::srv::ChangeState>(
+      node_change_state_topic);
+  }
+
+  unsigned int
+  get_state(std::chrono::seconds time_out = 3_s)
+  {
+    auto request = std::make_shared<lifecycle_msgs::srv::GetState::Request>();
+
+    if (!client_get_state_->wait_for_service(time_out)) {
+      fprintf(stderr, "Service %s is not available.\n",
+        client_get_state_->get_service_name().c_str());
+      return lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN;
+    }
+
+    auto result = client_get_state_->async_send_request(request);
+    if (result.wait_for(std::chrono::milliseconds(time_out)) != std::future_status::ready) {
+      fprintf(stderr, "[%s] Failed to get current state for node %s. Server timed out.\n",
+        get_name(), lifecycle_node);
+      return lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN;
+    }
+
+    if (result.get()) {
+      printf("[%s] Node %s has current state %s.\n",
+        get_name(), lifecycle_node, result.get()->current_state.label.c_str());
+      return result.get()->current_state.id;
+    } else {
+      fprintf(stderr, "[%s] Failed to get current state for node %s\n",
+        get_name(), lifecycle_node);
+      return lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN;
+    }
+  }
+
+  bool
+  change_state(std::uint8_t transition, std::chrono::seconds time_out = 3_s)
+  {
+    auto request = std::make_shared<lifecycle_msgs::srv::ChangeState::Request>();
+    request->transition.id = transition;
+
+    if (!client_change_state_->wait_for_service(time_out)) {
+      fprintf(stderr, "Service %s is not available.\n",
+        client_change_state_->get_service_name().c_str());
+      return false;
+    }
+
+    auto result = client_change_state_->async_send_request(request);
+    if (result.wait_for(std::chrono::milliseconds(time_out)) != std::future_status::ready) {
```

Comment:
No we don't, because we do not have a way to block on a service call while spin is happening in a different thread or from within a spin thread.

You could modify this to check `rclcpp::ok()` periodically:

```c++
auto end = std::chrono::steady_clock::now() + std::chrono::milliseconds(time_out);
std::chrono::milliseconds wait_period(100);
std::future_status status;
do {
  auto now = std::chrono::steady_clock::now();
  auto time_left = end - now;
  if (time_left <= 0) {break;}
  status = result.wait_for((time_left < wait_period) ? time_left : wait_period);
} while (rclcpp::ok() && status != std::future_status::ready)

if (status != std::future_status::ready) {
  // ... error handling
}
```

It's a bit tedious, but it accomplishes the goal. You could imagine generalizing this into a function in rclcpp, which a signature like this:

```c++
// throw exception when rclcpp::ok() != true
template<typename FutureT, typename WaitTimeT, typename WaitPeriodT = std::chrono::milliseconds>
std::future_status
rclcpp::wait_for_future(
  FutureT & future,
  WaitTimeT time_to_wait,
  WaitPeriodT max_wait_period = std::chrono::milliseconds(100));
```